Remove unnecessary checks when exporting gdextension binaries and allow using a prefix to auto-detect files

groud opened the pull request:
Closes https://github.com/godotengine/godot/issues/63230
May help with https://github.com/godotengine/godot/issues/53958, if it's not fixed already. I don't know.

This is something we discussed with @bruvzg and @Faless, it does three things:
- Remove a constraint that libs with relative paths were not copied into an exported project.
- Allow specifying an `autodetect_library_prefix` that allows automatically finding files matching the given export features. The logic simply search for the file with the higher number of matching tags.
- Likely fixes an unreported bug of `[dependencies]` section in the `.gdextension` config file not being processed at all.

Also, to avoid duplicated code between the exporter and the loader, I refactored the code to have a single function looking for the library file.

groud committed:
Remove uncessecary checks when exporting gdextension binaries

groud committed:
Allow specifying a prefix to automatically detect library files for gdextension exports

bruvzg commented on editor/plugins/gdextension_export_plugin.h:
```diff
@@ -102,6 +106,82 @@ void GDExtensionExportPlugin::_export_file(const String &p_path, const String &p
+						file_tags.append(tag);
+					}
+				}
+				if (file_tags.size() > best_file_tags.size()) {
```
It should unconditionally reject files with any tag that is not in `p_features`.

Or something like `libname.wrong_os.release.standalone.etc.etc2.wrong_arc` will have higher priority (4 matching tags: `release`, `standalone`, `etc`, `etc2`) than `libname.correct_os.release.correct_arch` (3 matching tags).

groud replied:
The problem is identifying parts of the filename that are tags and those that are not. 
I suppose a solution might be to simply consider that everything after the prefixes are tags besides the extension, but that mean users won't be able to have any custom content after the prefix.

bruvzg replied:
Restricting a file name to allow only tags and extension after the prefix seems reasonable. Or we can have a full list of supported tags (all archs and platforms).

groud replied:
Ah yeah, with a full list of arch and platforms it could work, but I wonder if users can define their one features. In such case I am not sure it would be possible, unless those features can be listed and transformed into identifiable tags 🤔

Faless commented on editor/plugins/gdextension_export_plugin.h:
```diff
@@ -102,6 +106,87 @@ void GDExtensionExportPlugin::_export_file(const String &p_path, const String &p
+		// Find the folder and file parts of the prefix.
+		String folder;
+		String file_prefix;
+		if (DirAccess::dir_exists_absolute(folder)) {
```
```suggestion
		if (DirAccess::dir_exists_absolute(autodetect_path)) {
```

Faless commented on editor/plugins/gdextension_export_plugin.h:
```diff
@@ -102,6 +106,87 @@ void GDExtensionExportPlugin::_export_file(const String &p_path, const String &p
+		String file_prefix;
+		if (DirAccess::dir_exists_absolute(folder)) {
+			folder = autodetect_path;
+		} else if (DirAccess::dir_exists_absolute(folder.get_base_dir())) {
```
```suggestion
		} else if (DirAccess::dir_exists_absolute(autodetect_path.get_base_dir())) {
```

Faless commented on editor/plugins/gdextension_export_plugin.h:
```diff
@@ -102,6 +106,87 @@ void GDExtensionExportPlugin::_export_file(const String &p_path, const String &p
+			String library_path = folder.path_join(best_file);
+			add_shared_object(library_path, best_file_tags);
+
+			if (p_features.has("iOS") && (library_path.ends_with(".a") || library_path.ends_with(".xcframework"))) {
+				String additional_code = "extern void register_dynamic_symbol(char *name, void *address);\n"
+										 "extern void add_ios_init_callback(void (*cb)());\n"
+										 "\n"
+										 "extern \"C\" void $ENTRY();\n"
+										 "void $ENTRY_init() {\n"
+										 "  if (&$ENTRY) register_dynamic_symbol((char *)\"$ENTRY\", (void *)$ENTRY);\n"
+										 "}\n"
+										 "struct $ENTRY_struct {\n"
+										 "  $ENTRY_struct() {\n"
+										 "    add_ios_init_callback($ENTRY_init);\n"
+										 "  }\n"
+										 "};\n"
+										 "$ENTRY_struct $ENTRY_struct_instance;\n\n";
+				additional_code = additional_code.replace("$ENTRY", entry_symbol);
+				add_ios_cpp_code(additional_code);
+
+				String linker_flags = "-Wl,-U,_" + entry_symbol;
+				add_ios_linker_flags(linker_flags);
+			}
+			could_export = true;
+		}
```
I this the same code as in the other for-loop? If yes, I'd suggest moving it to a dedicated method:

```
_add_ios_extension_code(entry)
```

I guess at some point it will have to become something like `_platform_add_extension(file, entry, tags)` so it can be properly implemented by the platform-specific exporter.

groud replied:
This code isn't duplicated anymore, now that I simply return the file path.

touilleMan commented on core/extension/native_extension.cpp:
```diff
@@ -39,6 +39,105 @@ String NativeExtension::get_extension_list_config_file() {
+		List<String> libraries;
+		p_config->get_section_keys("libraries", &libraries);
+
+		// First, check in the library list has an adequate file.
```
```c
// Second, ... 
```
?

DmitriySalnikov commented:
> * Likely fixes an unreported bug of `[dependencies]` section in the `.gdextension` config file not being processed at all.

Actually `dependencies` can be processed if you place it first, and then `libraries`.
The following config template works as intended in `beta6`, including exporting to android (the exported android app doesn't run for me yet, but the library is added to the apk correctly) v4.0.beta6.official [7f8ecffa5]
```
[configuration]

entry_symbol = "godot_init"

[dependencies]

macos.debug = [ ]
macos.release = [ ]
windows.debug.x86_64 = [ ]
windows.release.x86_64 = [ ]
linux.debug.x86_64 = [ ]
linux.release.x86_64 = [ ]

android.release.arm32 = [ ]
android.release.arm64 = [ ]
android.release.x86_32 = [ ]
android.release.x86_64 = [ ]

[libraries]

macos.debug = "res://..."
macos.release = "res://..."
windows.debug.x86_64 = "res://..."
windows.release.x86_64 = "res://..."
linux.debug.x86_64 = "res://..."
linux.release.x86_64 = "res://..."

android.release.arm32 = "res://..."
android.release.arm64 = "res://..."
android.release.x86_32 = "res://..."
android.release.x86_64 = "res://..."

# Repeat for other architectures to support arm64, rv64, etc.
```

But using the master branch (68e3f49157b10b87504a88fc31d15c4d1c691ead) with this PR (there was one conflict with error printing) breaks how tags are selected.
I got this error with the same config that works in beta6:
```
core\extension\native_extension.cpp:523 - No GDExtension library found for current OS and architecture (windows.x86_64) in configuration file: res://...
```

Part of the actual config:
```
[configuration]

entry_symbol = "godot_qoi_library_init"

[dependencies]

windows.debug.x86_64 = [ ]
windows.release.x86_64 = [ ]

[libraries]

windows.debug.x86_64 = "res://addons/qoi/libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.release.x86_64 = "res://addons/qoi/libs/win/godot_qoi.windows.template_release.x86_64.dll"
```

If I replace `windows.debug.x86_64` with `windows.x86_64`, then the library loads correctly in the editor. Also, when exporting a release, the same tag is used `windows.x86_64`, not `windows.release.x86_64`.

groud commented:
Oops, misclicked.

> Actually dependencies can be processed if you place it first, and then libraries.

If you have a look to the `GDExtensionExportPlugin::_export_file` function, the dependencies section is retrieved into `List<String> dependencies` but is then never used again. So I don't think files can be exported correctly. I am not sure if your examples are what you use exactly, but with empty arrays it's not possible to test if it works or not. Dependencies should be exported correctly, which I doubt they are without this PR.

> I got this error with the same config that works in beta6:

I think the error is expected. The logic is that all tags in `[libraries]` identifier should match the ones provided by the export. Here you have one tag more for each line, so it should not match. I think it is expected that you get an error, as the system cannot know which one to choose with not enough tags anyway.

If you did not mess your rebase, I think the fact exporting was working before may be due to the wrong code that was using `libraries` instead of `dependencies` in `_export_file`. But I am not 100% sure.

DmitriySalnikov commented:
> If you have a look to the `GDExtensionExportPlugin::_export_file` function, the dependencies section is retrieved into `List<String> dependencies` but is then never used again.

Yes, I've seen it. Unfortunately, dependencies probably don't work in beta6, but at least you can export with these lines.


> I think the error is expected. The logic is that all tags in `[libraries]` identifier should match the ones provided by the export. Here you have one tag more for each line, so it should not match. I think it is expected that you get an error, as the system cannot know which one to choose with not enough tags anyway.

> If I replace `windows.debug.x86_64` with `windows.x86_64`, then the library loads correctly in the editor. Also, when exporting a release, the same tag is used `windows.x86_64`, not `windows.release.x86_64`.

Now we have `editor`, `template_debug` and `template_release` and different architectures. `editor` and `template_debug` seem to be compatible, but `template_release` is not. And my release export still used `windows.x86_64`, not `windows.release.x86_64`. This means that the same library will try to load in all 3 configurations for `x86_64`.

In addition, the same format is used in the example: https://github.com/godotengine/godot-cpp/blob/576bd172851932ba3db95fda8760b4f297c89efd/test/demo/example.gdextension

But according to the logic of `ProjectSettings`/`ConfigFile`, this is the expected behavior, unfortunately.
In the 3.x version, you can add several tags, but they seem to work the same way as in 4.x. (only "64 string" will be used)
![image](https://user-images.githubusercontent.com/7782218/204536566-a6c43411-9289-4931-aaef-68e73467d2a9.png)

Maybe it's just worth improving the handling of `ConfigFile`s to support multiple tags?

akien-mga commented on core/extension/native_extension.h:
```diff
@@ -31,7 +31,10 @@
+#include <functional>
+
 #include "core/extension/gdnative_interface.h"
+#include "core/io/config_file.h"
```
Could possibly be forward declared? Not very important though.

groud reviewed (commented):
Approved in PR review meeting. Needs the "standalone" feature back, but renamed as "template".

akien-mga reviewed (approved):
Approved in GDExtension PR meeting.
We discussed that the `standalone` feature defined for both types of templates should likely still be kept, but likely renamed `template` (so e.g. on a template_debug build you have both `template_debug` and `template` defined).

DmitriySalnikov commented:
Will it be possible to specify in `.gdextension` `debug/editor` and `release` builds?

akien-mga commented:
> Will it be possible to specify in `.gdextension` `debug/editor` and `release` builds?

I think not, that's a valid concern indeed. It's a valid use case IMO to use the same library build for `editor` and for `template_debug`, which prior to this PR would have been matched by just using `debug`.

So maybe we should define all these:
```
# Fully-specific build configs:
editor
template_debug
template_release

# Actual features shared between builds
debug
release  (well only valid for `template_release` but for consistency)
template
```

So:
- `target=editor` would match `editor`, `debug`
- `target=template_debug` would match `template_debug`, `template`, `debug`
- `target=template_release` would match `template_release`, `template`, `release`

akien-mga commented:
> Maybe I should implement the same "return the most matching lib" behavior I implemented for the autodetection ?

Yeah that sounds good.

But here's a potential problem:
```
[libraries]

windows.debug = "bin/windows.debug.dll"
windows.editor = "bin/windows.editor.dll"
```
which one should be chosen?

We might need something a bit more advanced than just OS features for this if we want to differentiate between the best match and other suitable matches.

groud commented:
> Which one should be chosen? Should it raise an error?

If you only have neither "debug" or "editor", then yeah, it should fail. All tags in the identifier should be in the provided feature list for it to be selected. This is what the current behavior already implement.

The only problem is that in only matches the first match right now, while I think the "most precise" should likely be selected instead.

akien-mga commented:
Yeah but my question is: Which one will be considered the "most precise" for an editor build between `editor` and `debug`?

DmitriySalnikov commented:
Just checked the latest build and found a couple of problems. v4.0.beta.custom_build [181687847]

First I launched the editor without changing the tags in `.gdextension` and it started without any problems.
Here is the contents of the file
```ini
[configuration]

entry_symbol = "godot_qoi_library_init"

[libraries]

windows.debug.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.release.x86_64 = "libs/win/godot_qoi.windows.template_release.x86_64.dll"
```

But when exporting, there was an error related to the formatting of the error string
```
ERROR: not all arguments converted during string formatting
   at: (.\core/variant/variant.h:820)
ERROR: Method/function failed.
   at: GDExtensionExportPlugin::_export_file (.\editor/plugins/gdextension_export_plugin.h:88)
```

Next I added `template_` and the project was successfully exported and launched
```ini
[configuration]

entry_symbol = "godot_qoi_library_init"

[libraries]

windows.template_debug.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.template_release.x86_64 = "libs/win/godot_qoi.windows.template_release.x86_64.dll"
```

But after that, the editor can't find the right library. And I need to explicitly specify another line with `windows.editor.x86_64`. Actually, it's not a big problem, so it's probably even better that I need to explicitly specify all 3 configurations. So this is just for reference.
```ini
[configuration]

entry_symbol = "godot_qoi_library_init"

[libraries]

windows.editor.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.template_debug.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.template_release.x86_64 = "libs/win/godot_qoi.windows.template_release.x86_64.dll"
```

---

To check `autodetect_library_prefix` I removed `libs/` from the library paths and added `autodetect_library_prefix = "res://addons/qoi/libs/"`. In this case, the paths to the libraries can no longer be found relative to the `qoi.gdextension` file.
```ini
[configuration]

entry_symbol = "godot_qoi_library_init"
autodetect_library_prefix = "res://addons/qoi/libs/"

[libraries]

windows.editor.x86_64 = "win/godot_qoi.windows.editor.x86_64.dll"
windows.template_debug.x86_64 = "win/godot_qoi.windows.editor.x86_64.dll"
windows.template_release.x86_64 = "win/godot_qoi.windows.template_release.x86_64.dll"
```
```
│   .gitattributes
│   .gitignore
│   export_presets.cfg
│   icon.svg
│   icon.svg.import
│   project.godot
│
├───addons
│   └───qoi
│       │   LICENSE
│       │   qoi.gdextension
│       │   README.md
│       │
│       └───libs
│           ├───android
│           │       libgodot_qoi.android.template_debug.arm32.so
│           │       libgodot_qoi.android.template_debug.arm64.so
│           │       libgodot_qoi.android.template_debug.x86_32.so
│           │       libgodot_qoi.android.template_debug.x86_64.so
│           │       libgodot_qoi.android.template_release.arm32.so
│           │       libgodot_qoi.android.template_release.arm64.so
│           │       libgodot_qoi.android.template_release.x86_32.so
│           │       libgodot_qoi.android.template_release.x86_64.so
│           │
│           ├───linux
│           │       libgodot_qoi.linux.editor.x86_64.so
│           │       libgodot_qoi.linux.template_release.x86_64.so
│           │
│           ├───macos
│           │       libgodot_qoi.macos.editor.universal.dylib
│           │       libgodot_qoi.macos.template_release.universal.dylib
│           │
│           └───win
│                   godot_qoi.windows.editor.x86_64.dll
│                   godot_qoi.windows.template_release.x86_64.dll
|....
```

But the editor and the project export could not find these libraries. It looks like `autodetect_library_prefix` is not used at all.
```
ERROR: Failed to open C:/My/Projects/GE/4444/addons/qoi/win/godot_qoi.windows.template_release.x86_64.dll
   at: (core\io\dir_access.cpp:346)
```

Even using completely wrong paths like `autodetect_library_prefix = "libssss"` I don't get errors related to `autodetect_library_prefix`
```
ERROR: Can't open dynamic library: C:/My/Projects/GE/4444/addons/qoi/win/godot_qoi.windows.editor.x86_64.dll, error: Error 126: ...
   at: (platform\windows\os_windows.cpp:256)
ERROR: GDExtension dynamic library not found: C:/My/Projects/GE/4444/addons/qoi/win/godot_qoi.windows.editor.x86_64.dll
   at: NativeExtension::open_library (core\extension\native_extension.cpp:400)
ERROR: Failed loading resource: res://addons/qoi/qoi.gdextension. Make sure resources have been imported by opening the.   at: (core\io\resource_loader.cpp:215)
```

---

I also tried adding `dependencies`, but the specified files were not copied next to the `exe`, and there were no error messages.
```ini
[configuration]

entry_symbol = "godot_qoi_library_init"

[dependencies]

windows.template_debug.x86_64 = ["res://icon.svg", "res://shared_object.dll"]
windows.template_release.x86_64 = ["res://icon.svg", "res://shared_object.dll"]

[libraries]

windows.editor.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.template_debug.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.template_release.x86_64 = "libs/win/godot_qoi.windows.template_release.x86_64.dll"
```

groud commented:
> But when exporting, there was an error related to the formatting of the error string

That's weird, but I don't think it's a related to this PR. It might be worth opening a dedicated issue.

>But after that, the editor can't find the right library, and I had to explicitly specify another line with windows.editor.x86_64. Actually, it's not a big problem, so it's probably even better that I need to explicitly specify all 3 configurations. 

Technically, you don't need to, this should work:
```
[configuration]

entry_symbol = "godot_qoi_library_init"

[libraries]

windows.x86_64 = "libs/win/godot_qoi.windows.editor.x86_64.dll"
windows.template_release.x86_64 = "libs/win/godot_qoi.windows.template_release.x86_64.dll"
```
The algorithm should now select the most precise line out of the two, when all fits. Otherwise it will fall back to the first line.

> To check autodetect_library_prefix I removed libs/ from the library paths and added autodetect_library_prefix = "res://addons/qoi/libs/". In this case, the paths to the libraries can no longer be found relative to the qoi.gdextension file.

The prefix should include also the files prefix. A prefix is not a path to a folder, it MUST include the part of the file that does not consist of acceptable tags. Your library prefix should then be:
```
autodetect_library_prefix = "res://addons/qoi/libs/libgodot_qoi"
```
This is provided you put all library files in the exact same folder. You should not use subfolders there.

groud commented:
> What about dependencies?

I don't know. maybe try to open an issue with a MRP maybe.  This was not really the original goal of this PR, so it might deserve another one.

DmitriySalnikov commented:
Thanks. I'll open new issues if needed.

bruvzg commented:
> I also tried adding dependencies, but the specified files were not copied next to the exe, and there were no error messages.

I think dependencies are expected to be in a `Dictonary` form to allow setting target location (relevant for macOS, non-executable files should be stored in a different location, it was added to allow embedding JRE). Have not tested if it's working, so it might be still broken:

```
[dependencies]

macos.release = [
    { "res://source/non_exefile.cfg" : "/Content/Resources/" },
    { "res://source/exe_file.dylib" : "/Content/Frameworks/" }
]
```

DmitriySalnikov commented:
> I think dependencies are expected to be in a `Dictonary` form to allow setting target location

@bruvzg Yes, it requires a `Dictionary`, but without embedding into an `Array`
Here is a working example:
```ini
[dependencies]

windows.x86_64 = {
    "res://icon.svg" : "Content/",
    "res://test.gd" : "",
    "README.md" : "",
}
# "README.md" is a relative path to .gdextension
```
This is what the folder looks like after exporting
![image](https://user-images.githubusercontent.com/7782218/206032461-4348b199-a185-43fe-ba92-dd7c3640208e.png)

Thanks for the tip!
